test/tbb/test_collaborative_call_once.cpp

```diff
@@ -0,0 +1,372 @@
+    // // this allows us to make sure all the calls in the outer loop get stuck (if they will) before the inner loop begins
+    // const auto time_initial_delay = std::chrono::milliseconds(100);
+
+    // auto busy_wait = [](std::chrono::microseconds time) {
+    //     auto start_t = std::chrono::steady_clock::now();
+    //     while (std::chrono::steady_clock::now() < start_t + time) {
+    //         // busy wait
+    //     }
+    // };
+    auto work = [&]() {
+        call_ct++;
+        // busy_wait(time_initial_delay);
+        utils::doDummyWork(100000);
+        tbb::parallel_for(tbb::blocked_range<size_t>(0, num_work_pieces, 1), [&](const tbb::blocked_range<size_t>& range) {
+            for (size_t i = range.begin(); i != range.end(); ++i) {
+                // busy_wait(time_per_work_piece);
+                utils::doDummyWork(100000);
+            }
+            std::lock_guard<std::mutex> lock(mutex);
+            int id = tbb::detail::d1::current_thread_index();
+            id_ct_map[id] += range.size();
+        });
+    };
+    // check that a bunch of the threads did some of the work
+    auto isParallel = [](const std::map<int, int> cts) {
+        return std::count_if(cts.begin(), cts.end(), [](const std::pair<int, int>& it) { return it.second >= 2; }) >= tbb::this_task_arena::max_concurrency() / 3;
+    };
+    {
+        id_ct_map.clear();
+        call_ct = 0;
+
+        call_once_threads(work);
+
+        REQUIRE(call_ct == 1);
+        REQUIRE(isParallel(id_ct_map));
+    }
+    {
+        id_ct_map.clear();
+        call_ct = 0;
+
+        call_once_in_parallel_for(work);
+
+        REQUIRE(call_ct == 1);
+        REQUIRE(isParallel(id_ct_map));
+    }
+}
+
+std::uint64_t naive_recursive_fib(int n) {
+    if (n <= 1) {
+        return 1;
+    }
+    std::uint64_t a, b;
+    tbb::parallel_invoke([&] { a = naive_recursive_fib(n - 2); },
+                         [&] { b = naive_recursive_fib(n - 1); });
+    return a + b;
+}
+
+using FibBuffer = std::vector<std::pair<tbb::collaborative_once_flag, std::uint64_t>>;
+std::uint64_t collaborative_recursive_fib(int n, FibBuffer& buffer) {
+    if (n <= 1) {
+        return 1;
+    }
+    tbb::collaborative_call_once(buffer[n].first, [&]() {
+        std::uint64_t a, b;
+        tbb::parallel_invoke([&] { a = collaborative_recursive_fib(n - 2, buffer); },
+                             [&] { b = collaborative_recursive_fib(n - 1, buffer); });
+        buffer[n].second = a + b;
+    });
+    return buffer[n].second;
+}
+
+std::uint64_t collaborative_recursive_fib(int n) {
+    FibBuffer buffer(n+1);
+    return collaborative_recursive_fib(n, buffer);
+}
+
+//! Correctness test for Fibonacci example
+//! \brief \ref interface \ref requirement
+TEST_CASE("fibonacci example") {
+    constexpr int N = 25;
+
+    auto naive = naive_recursive_fib(N);
+
+    auto collaborative = collaborative_recursive_fib(N);
+
+    REQUIRE(naive == collaborative);
```

Comment:
Collaborative Fib has O(N) complexity, so does it make sense to use big N? E.g., 93 and compare the result with `12200160415121876738`?